File: python/medium/209.minimum_size_subarray_sum/minimum_size_subarray_sum_target.py
from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        if len(nums) == 1:
            return int(nums[0] >= target)

        if nums[0] >= target:
            return 1

        min_len = float("inf")
        left = 0
        right = 0
        window_sum = nums[0]

        while left < len(nums):
            if window_sum < target and right + 1 == len(nums):
                break

            if window_sum < target:
                right += 1
                window_sum += nums[right]
                continue

            current_len = right + 1 - left
            if current_len < min_len:
                min_len = current_len
                if min_len == 1:
                    return 1

            window_sum -= nums[left]
            left += 1

        return min_len if min_len != float("inf") else 0

File: c++/medium/209.minimum_size_subarray_sum/minimum_size_subarray_sum_target.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        if(nums.size() == 1) return nums[0] >= target;

        if(nums[0] >= target) return 1;

        int minLen = INT_MAX;
        int left = 0;
        int right = 0;
        int windowSum = nums[0];

        while(left < nums.size()) {
            if (windowSum < target && right+1 == nums.size()) break;

            if(windowSum < target) {
                right++;
                windowSum += nums[right];
                continue;
            }

            int currentLen = right+1 - left;
            if(currentLen < minLen) {
                minLen = currentLen;
                if(minLen == 1) return minLen;
            }

            windowSum -= nums[left];
            left++;
        }

        return minLen < INT_MAX ? minLen : 0;
    }
};
